#!/bin/bash

# This bash file performs variant calling for Illumina paired end sequencing data, few steps also specify variations we need to apply if we use nanopore sequencing data

num_args=$#
# If it is paired end, the code will take 3 inputs; 2 fastq files and 1 reference genome file

if [ $num_args -eq 3 ]; then
  
  fastq_file_1=$1 #input 1st fastq file
  fastq_file_2=$2 #input 2nd fastq file
  ref_file=$3 # input reference genome
  
  # Extracting the sample name from the file
  sample=$(basename "$fastq_file_1" | sed 's/_1.fastq.gz//; s/_1.fastq//')
  # make a directory of the sample names to store the output files
  mkdir ${sample} 
 
  # index the reference genome
  bwa index ref.fasta

  # Quality control: The primary step is to check the quality of the reads before preceeding to any downstream analysis as the quality of the reads affects downstream analysis
  # Using fastqc 
  fastqc ${sample}_1.fastq.gz ${sample}_2.fastq.gz -o ${sample} 
  # output are directed to the directory created in the above command

  # Removing low quality reads, adapter seuqnces, polyG tail etc.
  # Fastp is tool which removes low quality reads and perform sother operations like polyG tail trimming, adapter trimming. Low quality reads, adapters used during library preparation, can affect the accuracy of downstream analysis so need to be removed 
  fastp -i ${sample}_1.fastq.gz -I ${sample}_2.fastq.gz -o ${sample}_trimmed_read1.fastq.gz -O ${sample}_trimmed_read2.fastq.gz --trim_front1 30 --trim_tail1 30 --trim_front2 30 --trim_tail2 30  -c   
  # removing 30 bp from start and end for both reads , -c denotes  enabling base correction by overlap analysis in paired-end mode.
  
  # Alignment to reference genome
  # There are many tools available for alignment, and the appropriate tool can be chosen based on factors such as read length and reference genome size.
  # Here BWA is used which works well for Illumina short reads 
  #SAM file generation using minimap2(the reference fasta file is the chosen ref file after manual inspection of the coverage plot)
  bwa mem ref.fasta -t 16 ${sample}_trimmed_read1.fastq.gz ${sample}_trimmed_read2.fastq.gz > ${sample}.sam 
  # For nanopore sequencing data, minimap2 is used for the alignment step. Minimpa2 is used to handle the long reads generated by Nanopore seuqncing data.  

  # convert SAM to BAM
  samtools view -bS ${sample}_ref.sam -o ${sample}_ref.bam
  
  #  sort the BAM file
  samtools sort ${sample}_ref.bam -o ${sample}_sorted_ref.bam 

  # getting coverage using samtools coverage
  samtools coverage ${sample}_sorted_ref.bam -o ./${sample}/${sample}_coverage.txt

  # getting variants using GATK
  gatk HaplotypeCaller -R covidref.fasta -I ${sample}_sorted_ref.bam -O ${sample}.vcf.gz 

else
  fastq_file_1=$1 #input 1st fastq file
  ref_file=$3 # input reference genome
  # single end data will take only 2 inputs (1 read file and 1 refeence genome file)
  
  # Extracting the sample name from the file
  sample=$(basename "$fastq_file_1" | sed 's/_1.fastq.gz//; s/_1.fastq//')
  # make a directory of the sample names to store the output files
  mkdir ${sample} 
 
  # index the reference genome
  bwa index ref.fasta

  # Quality control: The primary step is to check the quality of the reads before preceeding to any downstream analysis as the quality of the reads affects downstream analysis
  # Using fastqc 
  fastqc ${sample}_1.fastq.gz ${sample}_2.fastq.gz -o ${sample} 
  # output are directed to the directory created in the above command

  # Removing low quality reads, adapter seuqnces, polyG tail etc.
  # Fastp is tool which removes low quality reads and perform sother operations like polyG tail trimming, adapter trimming. Low quality reads, adapters used during library preparation, can affect the accuracy of downstream analysis so need to be removed 
  fastp -i ${sample}_1.fastq.gz -I ${sample}_2.fastq.gz -o ${sample}_trimmed_read1.fastq.gz -O ${sample}_trimmed_read2.fastq.gz --trim_front1 30 --trim_tail1 30 --trim_front2 30 --trim_tail2 30  -c   
  # removing 30 bp from start and end for both reads , -c denotes  enabling base correction by overlap analysis in paired-end mode.
  
  # Alignment to reference genome
  # There are many tools available for alignment, and the appropriate tool can be chosen based on factors such as read length and reference genome size.
  # Here BWA is used which works well for Illumina short reads 
  #SAM file generation using minimap2(the reference fasta file is the chosen ref file after manual inspection of the coverage plot)
  bwa mem ref.fasta -t 16 ${sample}_trimmed_read1.fastq.gz ${sample}_trimmed_read2.fastq.gz > ${sample}.sam 
  # For nanopore sequencing data, minimap2 is used for the alignment step. Minimpa2 is used to handle the long reads generated by Nanopore seuqncing data.  

  # convert SAM to BAM
  samtools view -bS ${sample}_ref.sam -o ${sample}_ref.bam
  
  #  sort the BAM file
  samtools sort ${sample}_ref.bam -o ${sample}_sorted_ref.bam 

  # getting coverage using samtools coverage
  samtools coverage ${sample}_sorted_ref.bam -o ./${sample}/${sample}_coverage.txt

  # getting variants using GATK
  gatk HaplotypeCaller -R covidref.fasta -I ${sample}_sorted_ref.bam -O ${sample}.vcf.gz

fi

 
